### PR TITLE
[feat] experimental MEVO layer

### PR DESCRIPTION
**Pull request by min-xu-ai**
## What does this PR do?

This introduces a custom NN layer called MemoryEfficientVocabOutput (MEVO). This layer includes a linear output projection and a crossentropy loss (softmax + NLL loss). The main goal for this custom layer is to support very large scale training with large batch size, long sequence length and large vocabulary size.

The techniques uses in this layer are the following:

1. custom fused, checkpointed, tiled functions that avoiding materializing a HUGE output matrix from the output projection.
2. custom backward to avoid multiple copies of gradient buffers.
3. custom FSDP wrapping to shard the parameter and gradients for the projection.

A benchmark and 2 unit tests are added to test this layer. The layer is compared with a baseline implementation for both correctness, memory and speed. The speed is largely reduced due to complex python pytorch operator sequences. In the future, we will address the speed issues. At this moment, due to reduced memory, we can increase the batch size for the overall model training and gain moderate speedup overall.

The associated FSDP change is already merged with #836. It might be useful to refer to that commit while reviewing.

Anyone is welcome to review and provide constructive feedback. Thanks!

This is experimental and once it is more validated and more ready for general consumption, I will add more extensive user doc.

## Before submitting

- [x] Did you have fun?
  - Make sure you had fun coding 🙃
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/main/CONTRIBUTING.md)?
- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
  - [x] N/A
- [ ] Did you make sure to update the docs?
  - [x] N/A
- [x] Did you write any new necessary tests?
  - [ ] N/A
- [x] Did you update the [changelog](https://github.com/facebookresearch/fairscale/blob/main/CHANGELOG.md)? (if needed)
  - [ ] N/A

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

**Comment by min-xu-ai** [comment]
This is ready for review. Please take a look whoever has the time. No pressure.

It is still experimental since I have yet to run end to end tests with loss curves. Unit tests do ensure numerical comparisons though. Also, speed is not great. More work is needed in the future, likely with Triton.

I will have a separate doc with detailed performance and memory analysis.

File: fairscale/experimental/nn/ssd_offload.py
Comment (min-xu-ai) [inline fairscale/experimental/nn/ssd_offload.py:169]: these whilespace change is added by pre-commit hook.

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py
Comment (min-xu-ai) [inline fairscale/nn/data_parallel/fully_sharded_data_parallel.py:1415]: this single line is for better debugging output.

File: fairscale/experimental/nn/mevo.py
Comment (blefaudeux) [inline fairscale/experimental/nn/mevo.py:168]: this is to save some memory/speed, because autograd is wrong there ? curious as of why this is needed basically

File: fairscale/experimental/nn/mevo.py
Comment (blefaudeux) [inline fairscale/experimental/nn/mevo.py:187]: the sequence here (detach, FW/BW with autograd, then no_grad again) is pretty confusing without much context. Could you add some explanations ? I'm a bit worried about future maintainability

File: fairscale/experimental/nn/mevo.py
Comment (blefaudeux) [inline fairscale/experimental/nn/mevo.py:430]: I don't get that, why not declare them directly as inputs and get some help from mypy ?

Comment (min-xu-ai) [reply fairscale/experimental/nn/mevo.py:430]: thanks for the review, Ben!

if I recall this correctly, the prototype for `forward` in mypy is actually the Any, Any one. That's why I used that to make mypy happier. If you know a way that I can use separate args more directly here without too much mypy magic, that'd be great!

File: fairscale/experimental/nn/mevo.py
Comment (min-xu-ai) [reply fairscale/experimental/nn/mevo.py:187]: Yes, very true! I will add more comments in the code around this.

File: fairscale/experimental/nn/mevo.py
Comment (min-xu-ai) [reply fairscale/experimental/nn/mevo.py:168]: This is for saving memory with checkpointing. I will have an extensive doc that describes the algorithm in more details. Will add more comment here too.

File: fairscale/experimental/nn/mevo.py
Comment (blefaudeux) [reply fairscale/experimental/nn/mevo.py:430]: oh I typically ignore the mypy warning on this line ( #type: ignore), because I think that declaring inputs helps with readability (a lot of people will look at the interface, a lot more than the number of people who will look at the code in any case). The issue with the inputs goes with Pytorch not being typed in general I think, so this #type: ignore is a little ugly but I can live with it

**Comment by min-xu-ai** [comment]
@blefaudeux, addressed your comments. Thanks for the review!

If others no major issues, I will merge pretty soon. That will unblock my work of integrating it with fairseq properly.

**Comment by anj-s** [comment]
> @blefaudeux, addressed your comments. Thanks for the review!
> 
> If others no major issues, I will merge pretty soon. That will unblock my work of integrating it with fairseq properly.



> @blefaudeux, addressed your comments. Thanks for the review!
> 
> If others no major issues, I will merge pretty soon. That will unblock my work of integrating it with fairseq properly.

Taking a look now, would be great if you could wait till tomorrow morning. thanks!

**Comment by min-xu-ai** [comment]
> Taking a look now, would be great if you could wait till tomorrow morning. thanks!

Thanks for taking a look! Of course!

File: fairscale/experimental/nn/mevo.py
Comment (anj-s) [inline fairscale/experimental/nn/mevo.py:78]: nit: I would say "use_log" instead of "log". makes it more clear that it is a bool.

File: fairscale/experimental/nn/mevo.py
Comment (anj-s) [inline fairscale/experimental/nn/mevo.py:130]: nit: early return in line 129?

File: fairscale/experimental/nn/mevo.py
Comment (anj-s) [inline fairscale/experimental/nn/mevo.py:134]: nit: missing period at the end of this sentence.

File: fairscale/experimental/nn/mevo.py
Comment (anj-s) [inline fairscale/experimental/nn/mevo.py:46]: will this be a 1-D or 2D target?

File: fairscale/experimental/nn/mevo.py
Comment (anj-s) [inline fairscale/experimental/nn/mevo.py:140]: what does fp stand for?

File: fairscale/experimental/nn/mevo.py
Comment (anj-s) [reply fairscale/experimental/nn/mevo.py:190]: thanks for adding the comments! they definitely help but I think we need a doc to be honest :) this is pretty complex and context is pretty low with just the code.

File: fairscale/experimental/nn/mevo.py
Comment (anj-s) [inline fairscale/experimental/nn/mevo.py:280]: can we rename i and w across this file to something that it represents? it might be easier to grok this code.

File: fairscale/experimental/nn/mevo.py
Comment (anj-s) [inline fairscale/experimental/nn/mevo.py:388]: just curious about why it is slower. do we know why?

File: fairscale/experimental/nn/mevo.py
Comment (anj-s) [inline fairscale/experimental/nn/mevo.py:123]: Why do we implement this? is it for testing convenience?

File: fairscale/experimental/nn/mevo.py
Comment (anj-s) [inline fairscale/experimental/nn/mevo.py:402]: does tile_factor have to be the same? how does it impact memory savings?

File: tests/nn/data_parallel/test_fsdp_shared_weights_mevo.py
Comment (anj-s) [inline tests/nn/data_parallel/test_fsdp_shared_weights_mevo.py:56]: MEVO should also work for non-shared layers right? Curious what the reason was for having that support.

**Comment by anj-s** [comment]
nice PR @min-xu-ai ! Would love to see a doc when you get a chance. Feel free to merge since my comments are more questions with a few nits. Without a doc I am not sure I can comment about the algorithm anyway. Thanks for waiting for the merge.

**Comment by min-xu-ai** [comment]
> nice PR @min-xu-ai ! Would love to see a doc when you get a chance. Feel free to merge since my comments are more questions with a few nits. Without a doc I am not sure I can comment about the algorithm anyway. Thanks for waiting for the merge.

Thank you so much for the review @anj-s! Yes, the doc is coming too. The comments and suggestions are great. Will address them!

File: tests/nn/data_parallel/test_fsdp_shared_weights_mevo.py
Comment (min-xu-ai) [reply tests/nn/data_parallel/test_fsdp_shared_weights_mevo.py:56]: The floor memory consumption won't be as good otherwise. If mevo is used without this sharding support, then its param and grad memory will stay unsharded (i.e. full) for the entire forward/backward pass. This reduces the memory usable by other layers. With this support, we shard them after the layer is done computing, leaving much more memory to other layers, like the attention layers.

File: fairscale/experimental/nn/mevo.py
Comment (min-xu-ai) [reply fairscale/experimental/nn/mevo.py:402]: It doesn't need to be the same. I haven't extensively tuning it. Will add a TODO.

File: fairscale/experimental/nn/mevo.py
Comment (min-xu-ai) [reply fairscale/experimental/nn/mevo.py:123]: correct. This is purely for testing and benchmarking.

Comment (min-xu-ai) [reply fairscale/experimental/nn/mevo.py:123]: added a comment.

File: fairscale/experimental/nn/mevo.py
Comment (min-xu-ai) [reply fairscale/experimental/nn/mevo.py:388]: Yes, we know some of it is due to re-computation. Others are likely due to python looping. This is definitely an important future work.

File: fairscale/experimental/nn/mevo.py
Comment (min-xu-ai) [reply fairscale/experimental/nn/mevo.py:140]: fp is full precision, sharing the same acronym with floating point. :-)

I will make those longer names.

File: fairscale/experimental/nn/mevo.py
Comment (min-xu-ai) [reply fairscale/experimental/nn/mevo.py:46]: target is always 1D in this layer. We just reshape the input when the input is 2D. input is also always reshaped from 3d to 2d.

File: fairscale/experimental/nn/mevo.py
Comment (min-xu-ai) [reply fairscale/experimental/nn/mevo.py:130]: sure.

File: fairscale/experimental/nn/mevo.py
Comment (min-xu-ai) [reply fairscale/experimental/nn/mevo.py:280]: Alright, I added an explanation of them at the first occurrence.

File: fairscale/experimental/nn/mevo.py
Comment (blefaudeux) [inline fairscale/experimental/nn/mevo.py:170]: <3

File: fairscale/experimental/nn/mevo.py
Comment (blefaudeux) [inline fairscale/experimental/nn/mevo.py:192]: it really helps, I never realized that custom backward functions could themselves build a new autograd graph, I always assumed (I was wrong) that autograd was hands off there.. thanks for the comments @min-xu-ai !

**Review by blefaudeux — APPROVED** [review]
Looks great to me, feels like some of that could be sped up either with the recent cuda graph optimization from pytorch or by lowering this into triton or cuda kernels, but step by step is a good approach I think !

**Comment by min-xu-ai** [comment]
> Looks great to me, feels like some of that could be sped up either with the recent cuda graph optimization from pytorch or by lowering this into triton or cuda kernels, but step by step is a good approach I think !

Couldn’t agree more!

**Pull request merged**
